src/mame/sega/sega_beena.cpp

```diff
@@ -71,131 +102,2074 @@
+
+	// Game NOR Flash ROM
+	// Some games have an unreferenced handler that jumps to 0xa0ffc000;
+	map(ROM_FLASH_BASE, ROM_FLASH_BASE + 0x7fffff).lr8(NAME([]() { return 0xff; })).mirror(0x800000);
+
+	// Direct bitmap with 2-byte BGR555 colors per pixel, same as tile palettes. Scaling also applied.
+	// The bitmap size is checked by games to be `w * h <= 0x63000`, so we allocate `0x63000 * 2 = 0xc6000`.
+	map(0xc0000000, 0xc00c5fff).ram().share("bitmap");
+	// Tilemap data for 2 layers (background and foreground).
+	map(0xc00c6000, 0xc00c68ff).ram().share("tilemap_scroll_x");
+	// Work RAM for tilemap data
+	map(0xc00c6900, 0xc00c7fff).ram();
+	map(0xc00c8000, 0xc00c9fff).ram().share("tilemap_fg");
+	map(0xc00ca000, 0xc00cbfff).ram().share("tilemap_bg");
+	// Work RAM for general use + stack (usually starts at 0xc00fff80)
+	map(0xc00cc000, 0xc00fffff).ram().share("workram");
+	// Tile data for 2 layers (background and foreground) + sprites
+	map(0xc0100000, 0xc017ffff).ram().rw(FUNC(sega_9h0_0008_state::tiles_layers_r), FUNC(sega_9h0_0008_state::tiles_layers_w));
+	map(0xc0180000, 0xc01fffff).ram().rw(FUNC(sega_9h0_0008_state::tiles_sprites_r), FUNC(sega_9h0_0008_state::tiles_sprites_w));
+}
+
+void sega_9h0_0008_state::update_sensors(offs_t offset)
+{
+	// Power status
+	// - bit 1 = 0: a "low battery" screen shows up;
+	// - bit 2 = 0: game code goes into infinite loop, maybe power button pressed?
+	if (offset == 0x38/4) {
+		m_io_sensor_regs[offset] = 3;
+	}
+}
+
+uint32_t sega_9h0_0008_state::io_sensors_r(offs_t offset)
+{
+	update_sensors(offset);
+
+	return m_io_sensor_regs[offset];
+}
+
+uint32_t sega_9h0_0008_state::video_reg_r(offs_t offset)
+{
+	return m_video_regs[offset];
+}
+
+void sega_9h0_0008_state::video_reg_w(offs_t offset, uint32_t data, uint32_t mem_mask)
+{
+	LOG("video_reg_w @ %08x: addr=%08x data=%08x mask=%08x\n", m_maincpu->pc(), offset * 4, data, mem_mask);
+
+	/*
+	   Video control
+	   - bit 1..4: rendering toggles
+	     - 0x1: enable rendering
+	     - 0x2: enable fg layer
+	     - 0x4: enable bg layer
+	   - bit 5..8: scroll data entries
+	     - 0x10: 0x900 for x-axis scanlines
+	     - 0x20:  0x90 for x-axis tiles
+	     - 0x40:  0xb4 for y-axis tiles
+	   - bit 9..12: video format
+	     - 0x100: vertical frequency: 60Hz (NTSC) if 0, 50Hz (PAL) if 1
+	     - 0x200: mode: 352x240 (scales 2x from top-left corner) if 0, 704x480 if 1
+	   - bit 13..16: direct bitmap render priority
+	     - 0x0000: before overlay sprites
+	     - 0x1000: before all sprites
+	     - 0x2000: before fg layer
+	     - 0x3000: before bg and fg layers
+	*/
+	if (offset == 0x10/4) {
+		m_scale = ((data & 0x200) == 0) ? 2 : 1;
+	}
+
+	COMBINE_DATA(&m_video_regs[offset]);
+}
+
+uint32_t sega_9h0_0008_state::midi_reg_r(offs_t offset)
+{
+	if (offset == 0 && m_midi_busy_count > 0) {
+		m_midi_busy_count--;
+		return 0x2;
+	}
+
+	return m_midi_regs[offset];
+}
```

Comment:
Same here – you need to check `machine().side_effects_disabled()` before causing side effects in a read handler.